### ISSUE TITLE
Fix `block-no-empty` false positives for `reportNeedlessDisables`

### ISSUE THREAD

**Issue by sergeh** [stylelint/stylelint#5388]
Needless disable is being reported for `block-no-empty` even though the rule is set to `true`. Removing the `stylelint-disable-line` will give an `Unexpected empty block` error

> Which rule, if any, is the bug related to?

`block-no-empty`

> What code is needed to reproduce the bug?

```css
li { /* stylelint-disable-line block-no-empty */
}

dl { /* stylelint-disable-line block-no-empty */
}
```

> What stylelint configuration is needed to reproduce the bug?

```json
{
  "reportNeedlessDisables": true,
  "rules": {
    "block-no-empty": true,
  }
}
```

> Which version of stylelint are you using?

`13.13.1`

> How are you running stylelint: CLI, PostCSS plugin, Node.js API?

`stylelint "./styles/**/*.[s]css"`

> What did you expect to happen?

No needless disables reported

> What actually happened (e.g. what warnings or errors did you get)?

```shell
test.css
7:5    Needless disable for "block-no-empty"
10:5    Needless disable for "block-no-empty"
```

**Comment by jeddy3** [comment]
@sergeh Thanks for the report and for using the template.

It looks like a bug, either with the positions in the `block-no-empty` rule or with the behaviour of the `/* stylelint-disable-line */` command.

I suspect the bug is caused by the stylelint command comment being inside the empty block as the following doesn't produce a violation:

```css
li { } /* stylelint-disable-line block-no-empty */
```

I've labelled the issue as ready to implement. Please consider [contributing](https://stylelint.io/contributing) if you have time.

There are [steps on how to fix a bug in a rule](https://stylelint.io/developer-guide/rules#fix-a-bug-in-a-rule) in the Developer guide.

**Comment by sergeh** [comment]
Thanks @jeddy3, I'll try to contribute in the next week or so!

**Comment by sergeh** [comment]
just a heads up that I get a similar positive for `no-empty-source` when I use either one of these disables

```
/* stylelint-disable */
/* stylelint-disable-line */
/* stylelint-disable no-empty-source */
```

Any hints on what file I should start looking into?  I'm assuming I should start with `lib/needlessDisables.js`

**Comment by jeddy3** [comment]
>  I'm assuming I should start with `lib/needlessDisables.js`

Originally, I thought it's the interplay between where the rules register the position of their violations and something in `lib/needlessDisables.js`. However, digging a bit deeper I realise both rules consider their respective things (source or block) _not_ empty if they contain a comment. So, the command comment itself negates the violation and is therefore reported as needless.

We should probably exclude command comments, rather than count them as not empty. Does that sound like the most intuitive behaviour? If so, we'll want to abstract the `isStylelintCommand` and merging of adjacent double-slashed comments stuff from [`assignDisabledRanges.js`](https://github.com/stylelint/stylelint/blob/master/lib/assignDisabledRanges.js) into a utility so that these rules can make use of them.

Alternatively, we better document the existing behaviour that any comment will remove the violation.

**Comment by fuhlig** [comment]
Also experiencing `needless-disables` for `declaration-block-no-duplicate-properties` and `scss/no-duplicate-dollar-variables`

```scss
.selector {
    width: 100%;
    width: calc(100% - #{($sass-var * 2)}); /* stylelint-disable-line declaration-block-no-duplicate-properties */
}
```

**Comment by jeddy3** [comment]
@fuhlig Please create [a new bug issue](https://github.com/stylelint/stylelint/issues/new?assignees=&labels=&template=REPORT_A_BUG.yml) as this issue is specifically about the `block-no-empty` rule.
